enterprise/server/remote_execution/filecache/filecache.go

```diff
@@ -442,6 +444,65 @@ func (c *fileCache) Write(ctx context.Context, node *repb.FileNode, b []byte) (n
 	return n, nil
 }
 
+type verifiedWriter struct {
+	ctx context.Context
+	fc  *fileCache
+
+	csum hash.Hash
+	node *repb.FileNode
+	file *os.File
+}
+
+func newVerifiedWriter(ctx context.Context, fc *fileCache, node *repb.FileNode, digestFunction repb.DigestFunction_Value, file *os.File) (*verifiedWriter, error) {
+	csum, err := digest.HashForDigestType(digestFunction)
+	if err != nil {
+		return nil, err
+	}
+	return &verifiedWriter{
+		ctx:  ctx,
+		fc:   fc,
+		csum: csum,
+		node: node,
+		file: file,
+	}, nil
+}
+
+func (v *verifiedWriter) Write(p []byte) (n int, err error) {
+	if _, err := v.csum.Write(p); err != nil {
+		return 0, err
+	}
+	return v.file.Write(p)
+}
+
+func (v *verifiedWriter) Commit() error {
+	hashStr := fmt.Sprintf("%x", v.csum.Sum(nil))
+	if v.node.GetDigest().GetHash() != hashStr {
+		return status.DataLossErrorf("data checksum %q does not match expected checksum %q", hashStr, v.node.GetDigest().GetHash())
+	}
+	return v.fc.AddFile(v.ctx, v.node, v.file.Name())
+}
+
+func (v *verifiedWriter) Close() error {
+	defer func() {
+		if err := os.Remove(v.file.Name()); err != nil {
+			log.Warningf("Failed to remove filecache temp file: %s", err)
+		}
+	}()
+	return v.file.Close()
```

Comment:
does it make sense to close the file in Commit so that no additional writes are possible once the file has been added to the cache?

Reply:
Done.